### DISCUSSION TITLE
How to add Custom Fonts to a Website using TailwindCss?

### DISCUSSION THREAD

**Discussion by ghost** [tailwindlabs/tailwindcss#2060] [Help] [answered]
Hey Guys. Im Pretty New to using Tailwind and trying to add custom fonts to HTML / CSS.

How exactly do i add a custom Font in my Website?

Greetings
Daniel K

**Comment by estevanmaito** [comment, marked as answer]
# How to add custom fonts to Tailwind CSS

I'll use Google fonts, but the process is the same for every approach.

Choose your approach to add the font to the project. In my case, I'll use a link in the `head` but you could use `@import`, `@font-face`, it doesn't matter, as long as it's available to be used.

```html
<link href="https://fonts.googleapis.com/css2?family=Inter&display=swap" rel="stylesheet"> 
```

Now, Tailwind needs to know that you want to use this font. To do this, there are two approaches: extending an existing `fontFamily` or creating a new one.

## Extending `fontFamily`

Here I'm going to add 'Inter' to the `sans-serif` stack, which is the dafault picked by browsers. Sans font styles in Tailwind live under the `font-sans` class and can be extended in this way:

```js
// tailwind.config.js
const defaultTheme = require('tailwindcss/defaultTheme')

module.exports = {
  purge: [],
  theme: {
    extend: {
      fontFamily: {
        sans: ['Inter', ...defaultTheme.fontFamily.sans],
      },
    },
  },
  variants: {},
  plugins: [],
}
```

This line does the trick:

```js
sans: ['Inter', ...defaultTheme.fontFamily.sans],
```

It is important to destructure the current value of `fontFamily.sans` so you still support the default stack as a fallback (not _required_, but a good practice).

## Creating a new font stack (utility)

In this approach, you will create a new utility alongside the default `font-sans`, `font-mono` and `font-serif` [see docs](https://tailwindcss.com/docs/font-family/).

```js
// tailwind.config.js
module.exports = {
  purge: [],
  theme: {
    fontFamily: {
      custom: ['Inter', 'sans-serif'],
    },
  },
  variants: {},
  plugins: [],
}
```

Note that we're not **extending** `theme` anymore. Doing this, the utility class `font-custom` will be available for you to use in your code.

With this last approach, you could still make use of the fallback as we did earlier, but I'll leave it to you.

Good reads:

- [Tailwind font family docs](https://tailwindcss.com/docs/font-family/)

**Reply by ghost** [reply]
Hey i run into a small Problem. If i resize my Browser Window, the Site jumps to Sans-Serif. Is it Because of my Browser (opera GX) or is there something else going on.
Thank you

**Reply by estevanmaito** [reply]
Could you create a small reproduction on https://codepen.io ?

Edit: or at least share some code here.

**Reply by ghost** [reply]
I uploaded a ZIP of the complete Project here: https://github.com/dlnkrg/cerebrum

HTML:
`<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
        <title>New page</title>
        <link rel="stylesheet" href="styles/output.css">
        <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.13.0/css/all.css">
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Black+Han+Sans&display=swap">
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Bowlby+One+SC&display=swap">
    </head>
    <body>
        <div class="bg-gray-800 p-4 w-full">
            
            <div class="p-2 text-gray-700 bg-gray-900 rounded-lg shadow-lg ">
                <span class="px-2 mr-2 border-r border-gray-800"> <img src="https://www.freepnglogos.com/uploads/spotify-logo-png/file-spotify-logo-png-4.png" alt="alt placeholder" class="w-8 h-8 -mt-1 inline mx-auto"> </span>
                <span class="px-1 hover:text-white cursor-pointer"> <i class="fas fa-stream p-2 bg-gray-800 rounded-full"> </i> </span>
                <span class="px-1 hover:text-white cursor-pointer"> <i class="fas fa-search p-2 bg-gray-800 rounded-full"> </i> </span>
                <span class="px-1 hover:text-white cursor-pointer"> <i class="fas fa-th p-2 bg-gray-800 rounded-full"> </i> </span>
            </div>
        </div>
        <div class=":mb-6 :mt-4 :top-0 bg-gray-800 flex items-center px-8 text-center md:px-12 lg:pb-5 lg:pt-4 lg:text-left">
            <div>
                <h1 class="md:text-4xl text-3xl text-indigo-600">Cerebrum</h1>
                <p class="mt-2 text-sm text-gray-500 md:text-base">We research Technologies to Connect People with Computers.</p>
                <div class="flex justify-center lg:justify-start mt-6">
                    <a class="px-4 py-3 bg-indigo-600 text-gray-200 text-xs font-semibold rounded hover:bg-gray-800" href="#">Support us</a>
                    <a class="mx-4 px-4 py-3 bg-gray-300 text-gray-900 text-xs font-semibold rounded hover:bg-gray-400" href="#">Learn More</a>
                </div>
            </div>
        </div>
    </body>
</html>`

Classes i added to output.css:
` @font-face {
      font-family: 'gilroylight';
      src: url('gilroy-light-webfont.woff2') format('woff2'),
           url('gilroy-light-webfont.woff') format('woff');
      font-weight: normal;
      font-style: normal;

  }




  @font-face {
      font-family: 'officinasanitcmediumbold';
      src: url('officina-san-itc-extra-bold-webfont.woff2') format('woff2'),
           url('officina-san-itc-extra-bold-webfont.woff') format('woff');
      font-weight: normal;
      font-style: normal;

  }




  @font-face {
      font-family: 'typo_round_bold_demoregular';
      src: url('typo_round_bold_demo-webfont.woff2') format('woff2'),
           url('typo_round_bold_demo-webfont.woff') format('woff');
      font-weight: normal;
      font-style: normal;

  }




  @font-face {
      font-family: 'typo_round_regular_demoRg';
      src: url('typo_round_regular_demo-webfont.woff2') format('woff2'),
           url('typo_round_regular_demo-webfont.woff') format('woff');
      font-weight: normal;
      font-style: normal;

  }`

My Tailwind config:
`module.exports = {
  purge: [],
  theme: {
    extend: { fontFamily: { custom: ['gilroyextrabold', 'sans-serif'],
    },
    },
  },
  variants: {},
  plugins: [],
}
`

**Reply by ghost** [reply]
Sorry Github seems to ignore the code attributes

**Reply by estevanmaito** [reply]
`custom` is not part of the Tailwind font stack, so you cannot `extend` it. Read the example I gave above under "Creating a new font stack (utility)" and [the docs](https://tailwindcss.com/docs/font-family/#font-families)

**Reply by NanoNulla** [reply]
If you want to include a custom local font, then this article is very useful. 

https://hamidafghan.me/blog/tailwind-css-add-fonts-form-your-local

**Comment by ghost** [comment]
Thank you very much, I already somewhat figured it out using fontface classes in CSS.

**Comment by harshamv** [comment]
Is this method still the right way to do it? 

In the official docs we dont place it under extend any more!

**Reply by simonswiss** [reply]
Hey!

Well, if you define a `fontFamily` object directly inside `theme`, whatever goes in it will completely override the entire set of font families.

For example, 

```js
const defaultTheme = require('tailwindcss/defaultTheme')

module.exports = {
  theme: {
    fontFamily: {
      sans: ['Comic Sans', ...defaultTheme.fontFamily.sans]
    }
  }
}
```

Will add `Comic Sans` to the `font-sans` stack, and **completely remove** `font-mono`, `font-serif` etc.

Doing the same inside of `extend`...

```diff
const defaultTheme = require('tailwindcss/defaultTheme')

module.exports = {
  theme: {
+    extend: {
      fontFamily: {
        sans: ['Comic Sans', ...defaultTheme.fontFamily.sans]
      }
+    }
  }
}
```

... will add `Comic Sans` to the `font-sans` stack, and **preserve** the other font families like `font-mono`, `font-serif` etc.

Hope it makes sense 👍

**Reply by harshamv** [reply]
Thanks it clear now.

Also what do you suggest in terms of importing the actual font files via google fonts. whats the best place to place it?

Right now Am importing it in above the Tailwind imports. or should i move that to the layout file?

**Comment by diegomurillocorrea** [comment]
Hello, I have a problem
I'm trying to add a font to my NextJS + TailwindCSS project, I have sat the fonts folder into de public folder. 
When I add the @layer and @font-face rules and then in the tailwind.config file I do not know what else I have to do
Can somebody help me please?

**Reply by simonswiss** [reply]
Hey @diegomurillocorrea!

Happy to help with this. In case you haven't seen it, I cover this exact topic in a YouTube video:

https://youtu.be/sOnBG2wUm1s?t=453

This is not using NextJS, but the process would be exactly the same for Next. If the fonts are not showing, it's likely an issue with the file paths in your `@font-face` declarations. Are you getting errors in the console about missing fonts?

If you can share your project on a public repo, happy to take a look for you 👍

**Reply by diegomurillocorrea** [reply]
Hi Simon, yes, in fact, yesterday I was looking for your youtube channel because there is where I have learned all what I know about tailwind. I could manage the problem, it was located in the format declaration. I was typing "ttf" and not "truetype". That was all, then I could get the fonts and everything is OK. Thank you for everything. You are a GOAT!!!😎 🐐 🥇

**Comment by NanoNulla** [comment]
If you want to add custom fonts from your local, this article might be very useful.
https://hamidafghan.me/blog/tailwind-css-add-fonts-form-your-local

**Comment by ferdianarid** [comment]
Define it in globals.css
```bash
@tailwind base;
@tailwind components;
@font-face {
    font-family: "gilroy-bold";
    src: local("GilroyBold"), url(../public/fonts/Gilroy-Bold.woff) format('woff');
    font-weight: 700;
}

@font-face {
    font-family: "gilroy-semibold";
    src: local("GilroySemibold"), url(../public/fonts/Gilroy-Semibold.woff) format('woff');
    font-weight: 600;
}
@tailwind utilities;
```

And then use in tailwind config js
```bash
module.exports = {
             theme: {
                     extend: {
                            fontFamily: {
                                   "gilroy-bold": ["gilroy-bold", "sans-serif"],
                                   "gilroy-semibold": ["gilroy-semibold", "sans-serif"],
                                   "gilroy-medium": ["gilroy-medium", "sans-serif"],
                                   "gilroy-normal": ["gilroy-normal", "sans-serif"],
                                  "gilroy-light": ["gilroy-light", "sans-serif"],
                                  "gilroy-extralight": ["gilroy-extralight", "sans-serif"],
                                  "gilroy-thin": ["gilroy-thin", "sans-serif"]
                            }
                    }
          }
}
```

FInally, use it in index.jsx or etc like this

```bash
<h1 className="font-gilroy-semibold">This is custom font gilroy semibold</h1>
```
